File: IPWorks SSH Samples/SFTP Client/php/sftp_client.php
<?php
/*
 * IPWorks SSH 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks SSH in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksssh
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworksssh_sftpclient.php');
require_once('../include/ipworksssh_const.php');
?>
<?php
class MySFTP extends IPWorksSSH_SFTPClient {
  function FireSSHServerAuthentication($param) {
    $param['accept'] = true;
    /*
      This demo uses the above setting to unconditionally accept the server's host key.
      In production you can set the SSHAcceptServerHostKeyFingerPrint configuration setting to the 16-byte MD5 fingerprint
      to explictly accept only one host key value (see Configuration section in the help file for more detail).
      You may also set the SSHAcceptServerHostKeyEncoded property to the public key file in PEM format
    */
    return $param;
  }
  function FireDirList($param) {
    echo $param['direntry'] . "\n";
  }
  function FireError($param) {
    echo $param['description'] . "\n";
  }
  function FireLog($param) {
    echo $param['message'] . "\n";
  }
}

if ($argc < 4) {
  echo "Usage: php sftp_client.php host user pass\n\n";
  echo "  host: the host to connect to\n";
  echo "  user: the username to use for authentication\n";
  echo "  pass: the password to use for authentication\n\n";
  echo "Example: php sftp_client.php 192.168.1.2 myusername mypassword\n";
  return;
} else {
  $remoteHost = $argv[1];
  $user = $argv[2];
  $password = $argv[3];
}

function input($prompt) {
  echo $prompt;
  $handle = fopen("php://stdin", "r");
  $data = trim(fgets($handle));
  fclose($handle);
  return $data;
}

try {
  $sftp = new MySFTP();
  $sftp->setSSHHost($remoteHost);
  $sftp->setSSHAuthMode(2); // password
  $sftp->setSSHUser($user);
  $sftp->setSSHPassword($password);

  // Default port for SFTP is 22
  $sftp->doSSHLogon($remoteHost, 22);
  echo "Type \"?\" for a list of commands.\n";

  while (true) {
    try {
      $sftp->setRemoteFile("");
      $data = input("sftp> ");
      $arguments = explode(" ", $data);
      $command = $arguments[0];

      if ($command == "?" || $command == "help") {
        echo "?       cd      ls      pwd\n";
        echo "get     put     rm      mv\n";
        echo "mkdir   rmdir   exit\n";
      } elseif ($command == "bye" || $command == "quit" || $command == "exit") {
        $sftp->doSSHLogoff();
        break;
      } elseif ($command == "cd") {
        if (count($arguments) > 1) {
          $sftp->doChangeRemotePath($arguments[1]);
        } else {
          echo "Usage: cd <path>\n";
        }
      } elseif ($command == "get") {
        // get <remotefile> [<localfile>]
        if (count($arguments) < 2) {
          $sftp->setRemoteFile(input("Remote File: "));
          $sftp->setLocalFile(input("Local File: "));
        } elseif (count($arguments) == 2) {
          $sftp->setRemoteFile($arguments[1]);
          $sftp->setLocalFile($arguments[1]);
        } else {
          $sftp->setRemoteFile($arguments[1]);
          $sftp->setLocalFile($arguments[2]);
        }
        $sftp->doDownload();
        echo "File downloaded\n";
      } elseif ($command == "put") {
        // put <localfile> <remotefile>
        if (count($arguments) < 2) {
          $sftp->setLocalFile(input("Local File: "));
          $sftp->setRemoteFile(input("Remote File: "));
        } elseif (count($arguments) == 3) {
          $sftp->setLocalFile($arguments[1]);
          $sftp->setRemoteFile($arguments[2]);
        } else {
          echo "Usage: put <localFile> <remoteFile>\n";
          continue;
        }
        $sftp->doUpload();
        echo "File uploaded\n";
      } elseif ($command == "ls") {
        if (count($arguments) > 1) {
          $pathname = $sftp->getRemotePath();
          $sftp->doChangeRemotePath($arguments[1]);
          $sftp->doListDirectory();
          $sftp->doChangeRemotePath($pathname);
        } else {
          $sftp->doListDirectory();
        }
      } elseif ($command == "pwd") {
        echo $sftp->getRemotePath() . "\n";
      } elseif ($command == "mkdir") {
        if (count($arguments) > 1) {
          $sftp->doMakeDirectory($arguments[1]);
        } else {
          echo "Usage: mkdir <directory>\n";
        }
      } elseif ($command == "rmdir") {
        if (count($arguments) > 1) {
          $sftp->doRemoveDirectory($arguments[1]);
        } else {
          echo "Usage: rmdir <directory>\n";
        }
      } elseif ($command == "rm" || $command == "delete") {
        if (count($arguments) > 1) {
          $sftp->doDeleteFile($arguments[1]);
        } else {
          echo "Usage: rm <remoteFile>\n";
        }
      } elseif ($command == "mv") {
        if (count($arguments) > 2) {
          $sftp->setRemoteFile($arguments[1]);
          $sftp->doRenameFile($arguments[2]);
        } else {
          echo "Usage: mv <filename> <newFileName>\n";
        }
      } elseif ($command == "") {
        // do nothing
      } else {
        echo "Bad command / Not implemented in demo.\n";
      }
    } catch (Exception $ex) {
      echo "Error: " . $ex->getMessage() . "\n";
    }
  }
} catch (Exception $e) {
  echo "Error: " . $e->getMessage() . "\n";
}

input("\npress <return> to continue...");

?>

File: IPWorks SSH Samples/SFTP Client/c++/sftpclient.cpp
/*
 * IPWorks SSH 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks SSH in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksssh
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworksssh.h"

#define LINE_LEN 80

class MySFTP : public SFTPClient
{
public:

	MySFTP()
	{
	}

	virtual int FireSSHServerAuthentication(SFTPClientSSHServerAuthenticationEventParams *e)
	{
		if (e->Accept) return 0;
		printf("\nServer provided the following fingerprint:\n %s\n",
		       e->Fingerprint);
		printf("Would you like to continue? [y/n] ");
		char command[LINE_LEN];
		fgets(command,LINE_LEN,stdin);
		command[strlen(command)-1] = '\0';
		if (!strcmp(command, "y")) e->Accept = true;
		else exit(0);
		return 0;
	}

	virtual int FireDirList(SFTPClientDirListEventParams *e)
	{
		printf( "%s\n", e->DirEntry );
		return 0;
	}

	virtual int FireSSHStatus(SFTPClientSSHStatusEventParams *e)
	{
		printf( "%s\n", e->Message );
		return 0;
	}

	virtual int FireError( SFTPClientErrorEventParams *e )
	{
		printf("Error %i: %s", e->ErrorCode, e->Description);
		return 0;
	}


};

int main(int argc, char **argv)
{

	MySFTP sftp;                  // sftp object
	char command[LINE_LEN];     // user's command
	char *argument;             // arguments to the user's command
	char pathname[LINE_LEN];    // for use with the ls command
	int ret_code=0;
	int port = 22;              // default port

	//  If at least three arguments follow "sftp" at the command line,
	//   read them and log the user into a server.
	if (argc >= 4)
	{
		sscanf(argv[2], "%d", &port);
		sftp.SetSSHUser(argv[3]);
		sftp.SetSSHPassword(argv[4]);
		ret_code = sftp.SSHLogon(argv[1], port);
	}
	else
	{
		printf ("SSH Server: " );
		fgets(command,LINE_LEN,stdin);
		command[strlen(command)-1] = '\0';
		sftp.SetSSHHost( command );
		printf("SSH Port: ");
		fgets(command, LINE_LEN, stdin);
		command[strlen(command) - 1] = '\0';
		sscanf(command, "%d", &port);
		sftp.SetSSHPort(port);
		printf ("User: " );
		fgets(command,LINE_LEN,stdin);
		command[strlen(command)-1] = '\0';
		sftp.SetSSHUser( command );
		printf("Password: ");
		fgets(command,LINE_LEN,stdin);
		command[strlen(command)-1] = '\0';
		sftp.SetSSHPassword( command );
		ret_code = sftp.SSHLogon(sftp.GetSSHHost(), sftp.GetSSHPort());
	}
	if (ret_code) goto done;

	while (1)
	{

		sftp.SetRemoteFile("");
		printf( "\nsftp> " );
		fgets(command,LINE_LEN,stdin);
		command[strlen(command)-1] = '\0';
		argument = strtok( command, " \t\n" );

		if ( ! strcmp(command, "?") )
		{
			printf( "?          exit      help     put\n"
			        "append     cd        ls       pwd\n"
			        "mkdir      rmdir     rm       get      mv\n");
		}

		else if ( ! strcmp(command, "append") )
		{
			argument = strtok( NULL, " \t\n" );
			sftp.SetLocalFile(argument);
			argument = strtok( NULL, " \t\n" );
			sftp.SetRemoteFile(argument);
			ret_code = sftp.Append();
		}

		else if ( ! strcmp(command, "exit") )
		{
			ret_code = sftp.SSHLogoff();
			exit(0);
		}

		else if ( ! strcmp(command, "cd") )
		{
			if ( argument = strtok( NULL, " \t\n" ) )
			{
				sftp.ChangeRemotePath(argument);
			}
		}

		else if ( ! strcmp(command, "get") )
		{
			argument = strtok( NULL, " \t\n" );
			sftp.SetRemoteFile(argument);
			sftp.SetLocalFile(argument);
			ret_code = sftp.Download();
			printf(  "Download complete.\n");
		}

		else if ( ! strcmp(command, "help") )
		{
			printf( "?          exit      help     put\n"
			        "append     cd        ls       pwd\n"
			        "mkdir      rmdir     rm       get      mv\n");
		}

		else if ( ! strcmp(command, "ls") )
		{
			if ( argument = strtok( NULL, " \t\n" ) )
			{
				strcpy(pathname, sftp.GetRemotePath());
				int ret_code = sftp.ChangeRemotePath(argument);
				if (!ret_code)
				{
					ret_code = sftp.ListDirectory();
				}
				if (!ret_code)
				{
					ret_code = sftp.ChangeRemotePath(pathname);
				}
			}
			else
			{
				ret_code = sftp.ListDirectory();
			}
		}

		else if ( ! strcmp(command, "mkdir") )
		{
			if ( argument = strtok( NULL, " \t\n" ) )
			{
				ret_code = sftp.MakeDirectory(argument);
			}
		}

		else if ( ! strcmp(command, "mv") )
		{
			argument = strtok( NULL, " \t\n" );
			sftp.SetRemoteFile(argument);
			argument = strtok( NULL, " \t\n" );
			ret_code = sftp.RenameFile(argument);
		}
		else if ( ! strcmp(command, "put") )
		{
			argument = strtok( NULL, " \t\n" );
			sftp.SetRemoteFile(argument);
			sftp.SetLocalFile(argument);
			ret_code = sftp.Upload();
			printf(  "Upload complete.\n");
		}
		else if ( ! strcmp(command, "pwd") )
		{
			printf( "%s\n", sftp.GetRemotePath() );
		}
		else if ( ! strcmp(command, "rm") )
		{
			if ( argument = strtok( NULL, " \t\n" ) )
			{
				ret_code = sftp.DeleteFile(argument);
			}
		}
		else if ( ! strcmp(command, "rmdir") )
		{
			if ( argument = strtok( NULL, " \t\n" ) )
			{
				ret_code = sftp.RemoveDirectory(argument);
			}
		}
		else if ( ! strcmp(command, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Bad command / Not implemented in demo.\n" );
		} // end of command checking
		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (sftp.GetLastError())
			{
				printf( " \"%s\"\n", sftp.GetLastError() );
			}
		}
		ret_code = 0;   // flush out error
	}  // end of main while loop

done:
	if (ret_code)     // Got an error.  The user is done.
	{
		printf( "\nError: %d", ret_code );
		if (sftp.GetLastError())
		{
			printf( " \"%s\"\n", sftp.GetLastError() );
		}
	}
	exit(ret_code);
	return 0;
}


